File: crates/plugins/tauri-plugin-appauth/android/src/main/java/app/tauri/appauth/ErrorMapping.kt
// SPDX-License-Identifier: Apache-2.0

package app.tauri.appauth

import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import net.openid.appauth.AuthorizationException

/// Maps AppAuth-Android `AuthorizationException`s onto the stable error codes
/// the Rust crate publishes via `Error::code()`.
///
/// AppAuth groups exceptions into four type categories that mirror the OAuth
/// spec's error surface (general / authorization / token / registration). The
/// `error` and `errorDescription` fields are surfaced verbatim to the JS layer
/// so callers don't have to parse free-form messages.
object ErrorMapping {

    const val CODE_USER_CANCELED = "USER_CANCELED"
    const val CODE_AUTHORIZATION_FAILED = "AUTHORIZATION_FAILED"
    const val CODE_TOKEN_EXCHANGE_FAILED = "TOKEN_EXCHANGE_FAILED"
    const val CODE_NETWORK_ERROR = "NETWORK_ERROR"
    const val CODE_INVALID_REGISTRATION_RESPONSE = "INVALID_REGISTRATION_RESPONSE"
    const val CODE_ID_TOKEN_VALIDATION_FAILED = "ID_TOKEN_VALIDATION_FAILED"
    const val CODE_BROWSER_NOT_AVAILABLE = "BROWSER_NOT_AVAILABLE"
    const val CODE_INVALID_REQUEST = "INVALID_REQUEST"
    const val CODE_SERVER_ERROR = "SERVER_ERROR"

    /// Reject `invoke` with the appropriate code/message/oauth-context combo.
    fun reject(invoke: Invoke, ex: AuthorizationException) {
        val code = when (ex.type) {
            AuthorizationException.TYPE_GENERAL_ERROR -> mapGeneral(ex)
            AuthorizationException.TYPE_OAUTH_AUTHORIZATION_ERROR -> mapAuthorization(ex)
            AuthorizationException.TYPE_OAUTH_TOKEN_ERROR -> CODE_TOKEN_EXCHANGE_FAILED
            AuthorizationException.TYPE_OAUTH_REGISTRATION_ERROR -> CODE_INVALID_REGISTRATION_RESPONSE
            AuthorizationException.TYPE_RESOURCE_SERVER_AUTHORIZATION_ERROR -> CODE_AUTHORIZATION_FAILED
            else -> CODE_AUTHORIZATION_FAILED
        }
        val message = ex.errorDescription ?: ex.error ?: ex.message ?: code
        val data = JSObject().apply {
            ex.error?.let { put("oauthError", it) }
            ex.errorDescription?.let { put("oauthErrorDescription", it) }
        }
        invoke.reject(message, code, ex, if (data.length() > 0) data else null)
    }

    /// `TYPE_GENERAL_ERROR` covers transport, browser, and validation failures.
    /// We split them into the per-cause buckets the JS layer needs.
    private fun mapGeneral(ex: AuthorizationException): String {
        return when (ex.code) {
            AuthorizationException.GeneralErrors.USER_CANCELED_AUTH_FLOW.code,
            AuthorizationException.GeneralErrors.PROGRAM_CANCELED_AUTH_FLOW.code -> CODE_USER_CANCELED

            AuthorizationException.GeneralErrors.NETWORK_ERROR.code -> CODE_NETWORK_ERROR

            AuthorizationException.GeneralErrors.SERVER_ERROR.code,
            AuthorizationException.GeneralErrors.INVALID_DISCOVERY_DOCUMENT.code,
            AuthorizationException.GeneralErrors.JSON_DESERIALIZATION_ERROR.code -> CODE_SERVER_ERROR

            AuthorizationException.GeneralErrors.ID_TOKEN_PARSING_ERROR.code,
            AuthorizationException.GeneralErrors.ID_TOKEN_VALIDATION_ERROR.code -> CODE_ID_TOKEN_VALIDATION_FAILED

            else -> CODE_AUTHORIZATION_FAILED
        }
    }

    private fun mapAuthorization(ex: AuthorizationException): String {
        // `access_denied` (USER_CANCELED equivalent at the OAuth level) is the
        // only authorization-endpoint error we promote out of AUTHORIZATION_FAILED
        // — every other code is a server-side denial that callers handle the
        // same way (display the OAuth error code from `data.oauthError`).
        return if (ex.error == "access_denied") CODE_USER_CANCELED else CODE_AUTHORIZATION_FAILED
    }
}

File: crates/plugins/tauri-plugin-appauth/ios/Sources/ErrorMapping.swift
// SPDX-License-Identifier: Apache-2.0

import AppAuth
import Foundation
import Tauri

/// Maps `NSError`s emitted by AppAuth (and by `URLSession` indirectly) onto the
/// stable error codes the Rust crate publishes via `Error::code()`.
///
/// AppAuth surfaces three OAuth-error sub-domains
/// (`OIDOAuthAuthorizationErrorDomain`, `OIDOAuthTokenErrorDomain`,
/// `OIDOAuthRegistrationErrorDomain`); for each we extract the OAuth `error`
/// and `error_description` strings from `userInfo[OIDOAuthErrorResponseErrorKey]`
/// so the JS layer doesn't have to parse free-form messages.
enum ErrorMapping {

    struct Mapping {
        let code: String
        let message: String
        let oauthError: String?
        let oauthErrorDescription: String?
    }

    /// Reject `invoke` with the appropriate code/message/oauth-context combo.
    static func reject(_ invoke: Invoke, error: Error) {
        let mapping = map(error)
        var data: JsonObject = [:]
        if let oauth = mapping.oauthError {
            data["oauthError"] = oauth
        }
        if let desc = mapping.oauthErrorDescription {
            data["oauthErrorDescription"] = desc
        }
        let payload: JsonValue? = data.isEmpty ? nil : .dictionary(data)
        invoke.reject(mapping.message, code: mapping.code, data: payload)
    }

    static func map(_ error: Error) -> Mapping {
        let nsError = error as NSError

        if nsError.domain == OIDOAuthAuthorizationErrorDomain {
            return mapOAuth(nsError, defaultCode: codeAuthorizationFailed)
        }
        if nsError.domain == OIDOAuthTokenErrorDomain {
            return mapOAuth(nsError, defaultCode: codeTokenExchangeFailed)
        }
        if nsError.domain == OIDOAuthRegistrationErrorDomain {
            return mapOAuth(nsError, defaultCode: codeInvalidRegistrationResponse)
        }
        if nsError.domain == OIDGeneralErrorDomain {
            return mapGeneral(nsError)
        }
        if nsError.domain == NSURLErrorDomain {
            return Mapping(
                code: codeNetworkError,
                message: nsError.localizedDescription,
                oauthError: nil,
                oauthErrorDescription: nil
            )
        }
        if let bridge = error as? AppAuthBridgeError {
            switch bridge {
            case .invalidRequest(let message):
                return Mapping(
                    code: codeInvalidRequest,
                    message: message,
                    oauthError: nil,
                    oauthErrorDescription: nil
                )
            case .serverError(let message):
                return Mapping(
                    code: codeServerError,
                    message: message,
                    oauthError: nil,
                    oauthErrorDescription: nil
                )
            }
        }
        return Mapping(
            code: codeAuthorizationFailed,
            message: nsError.localizedDescription,
            oauthError: nil,
            oauthErrorDescription: nil
        )
    }

    // MARK: - Sub-mappers

    /// `OIDGeneralErrorDomain` covers AppAuth's library-internal errors. The
    /// integer codes are stable per AppAuth's public API (see `OIDError.h`);
    /// we compare on `rawValue` rather than enum cases to sidestep Swift's
    /// inconsistent acronym-import rules for `OIDErrorCodeIDToken*`.
    private static func mapGeneral(_ nsError: NSError) -> Mapping {
        let message = nsError.localizedDescription
        switch nsError.code {
        case OIDErrorCode.userCanceledAuthorizationFlow.rawValue,
             OIDErrorCode.programCanceledAuthorizationFlow.rawValue:
            return Mapping(code: codeUserCanceled, message: message, oauthError: nil, oauthErrorDescription: nil)

        case OIDErrorCode.networkError.rawValue:
            return Mapping(code: codeNetworkError, message: message, oauthError: nil, oauthErrorDescription: nil)

        case OIDErrorCode.serverError.rawValue,
             OIDErrorCode.invalidDiscoveryDocument.rawValue,
             OIDErrorCode.jsonDeserializationError.rawValue,
             OIDErrorCode.jsonSerializationError.rawValue:
            return Mapping(code: codeServerError, message: message, oauthError: nil, oauthErrorDescription: nil)

        case OIDErrorCode.tokenResponseConstructionError.rawValue,
             OIDErrorCode.tokenRefreshError.rawValue:
            return Mapping(code: codeTokenExchangeFailed, message: message, oauthError: nil, oauthErrorDescription: nil)

        case OIDErrorCode.registrationResponseConstructionError.rawValue:
            return Mapping(code: codeInvalidRegistrationResponse, message: message, oauthError: nil, oauthErrorDescription: nil)

        case oidErrorCodeIDTokenParsing,
             oidErrorCodeIDTokenFailedValidation:
            return Mapping(code: codeIdTokenValidationFailed, message: message, oauthError: nil, oauthErrorDescription: nil)

        case OIDErrorCode.safariOpenError.rawValue,
             OIDErrorCode.browserOpenError.rawValue:
            return Mapping(code: codeBrowserNotAvailable, message: message, oauthError: nil, oauthErrorDescription: nil)

        default:
            return Mapping(code: codeAuthorizationFailed, message: message, oauthError: nil, oauthErrorDescription: nil)
        }
    }

    /// Pull `error` / `error_description` out of the OAuth response dict that
    /// AppAuth attaches to `userInfo[OIDOAuthErrorResponseErrorKey]`.
    private static func mapOAuth(_ nsError: NSError, defaultCode: String) -> Mapping {
        let response = nsError.userInfo[OIDOAuthErrorResponseErrorKey] as? [String: Any]
        let oauthError = response?[OIDOAuthErrorFieldError] as? String
        let oauthErrorDescription = response?[OIDOAuthErrorFieldErrorDescription] as? String
        return Mapping(
            code: defaultCode,
            message: nsError.localizedDescription,
            oauthError: oauthError,
            oauthErrorDescription: oauthErrorDescription
        )
    }

    // MARK: - Code constants (kept in sync with Rust `Error::code()`)

    static let codeUserCanceled = "USER_CANCELED"
    static let codeAuthorizationFailed = "AUTHORIZATION_FAILED"
    static let codeTokenExchangeFailed = "TOKEN_EXCHANGE_FAILED"
    static let codeNetworkError = "NETWORK_ERROR"
    static let codeInvalidRegistrationResponse = "INVALID_REGISTRATION_RESPONSE"
    static let codeIdTokenValidationFailed = "ID_TOKEN_VALIDATION_FAILED"
    static let codeBrowserNotAvailable = "BROWSER_NOT_AVAILABLE"
    static let codeInvalidRequest = "INVALID_REQUEST"
    static let codeServerError = "SERVER_ERROR"

    // ID Token error codes from `OIDError.h` (-14 / -15). Hard-coded because
    // Swift's import of `OIDErrorCodeIDTokenParsingError` produces a case name
    // (`iDTokenParsingError`) that is unstable across compiler versions.
    private static let oidErrorCodeIDTokenParsing: Int = -14
    private static let oidErrorCodeIDTokenFailedValidation: Int = -15
}
